rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+installation target, and the appropriate installation method. This resource may
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
+
+- `version-controller`: A pluggable component responsible for generating the
+`version-directive` based on some dynamic state (e.g. a server which publishes
+package versions and hashes). A builtin `version-controller` would simply be a
+configuration resource from the user's perspective. A custom/external version
+controller would be any program with the permissions necessary to update the
+`version-directive` resource.
+
+- *version reconciliation loop*: A control loop that runs in the auth server which
+compares the desired state as specified by the `version-directive` with the current
+state of teleport installations across the cluster. When mismatches are discovered,
+the appropriate `installer`s are run.
+
+- `installer`: A component capable of attempting to affect installation
+of a specific target on one or more teleport hosts. The auth server needs to know
+enough to at least start the process, but the core logic of a given installer
+may be an external command (e.g. the proposed `local-script` installer would cause
+each teleport instance in need of upgrade to run a user-supplied script locally and
+then restart). From the perspective of a user, an installer is a teleport
+configuration object (though that configuration object may only be a thin hook into
+the "real" `installer`). Whether or not the teleport instance being upgraded
+needs to understand the installer will vary depending on type.
+
+There is room in the above model for additional granularity (e.g. schedulers), but
+the listed components are a decent starting point.
+
+
+### Implementation Phases
+
+Implementation will be divided into a series of phases consisting of 1 or more
+separate PRs/releases each:
+
+- Setup: Changes required for inventory status/control model, but not necessarily
+specific to the upgrade system.
+
+- Notification-Only System: Optional phase intended to deliver value sooner
+at the cost of overall feature development time.
+
+- Script-Based Installs MVP: Early MVP supporting only manual control and
+simple script-based installers for non-auth instances.
+
+- TUF-Based System MVP: Fully functional but minimalistic upgrade system
+based on TUF (still excludes auth instances).
+
+- Stability & Polish: Additional necessary features (including auth installs
+and local rollbacks). Represents the point at which the core upgrade system
+can be considered "complete".
+
+- Extended Feature Set: A collection of nice-to-haves that we're pretty sure
+folks are going to want.
+
+See the [implementation plan](#implementation-plan) section for detailed breakdown
+of the required elements for each phase.
+
+
+## Details
+
+### Usage Scenarios
+
+
+#### Notification-Only Usecase
```

Comment:
@xinding33 can we add some mockups of how the UI notifications will look like for users? Similar to the banner we have with "assume roles?"